### PR TITLE
Experimental beam library for verifiable maps

### PR DESCRIPTION
**Pull request by mhutchinson**
Tests pass on go 1.14.X and 1.15 now that the shims have been generated.
This includes a brief README, but demo code showing the generation of
a map and synthesis/checking of inclusion proofs would really help.

### CODE REVIEW

File: experimental/batchmap/tilepb/tile.proto [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  // All non-empty leaves in this tile, sorted left-to-right.
+  repeated TileLeaf leaves = 3;
+}
+
+// TileLeaf is a leaf value of a Tile.
+// If it belongs to a leaf tile then this represents one of the values that the
+// map commits to. Otherwise, this leaf represents the root of the subtree in
+// the stratum below.
+message TileLeaf {
+  // The path from the root of the container Tile to this leaf.
```
Comment (AlCutter) [inline experimental/batchmap/tilepb/tile.proto:24]: this is the same as the index of this leaf within its tile? (worth adding a comment along those lines?)

File: experimental/batchmap/tilepb/tile.proto [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  bytes path = 1;
+  // The hash value being committed to.
+  bytes hash = 2;
+}
+
+// Entry is a single key/value to be committed to by the map.
+message Entry {
+  // The key that uniquely identifies this key/value.
+  // These keys must be distributed uniformly and randomly across the key space.
```
Comment (AlCutter) [inline experimental/batchmap/tilepb/tile.proto:33]: Maybe worth just connecting up the dots and explicitly saying we use a cryptographically secure hash of the application "key" to provide this property?

File: experimental/batchmap/tmap.go [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode leaf ID: %v", err)
+		}
+		nodes = append(nodes, smt.Node{ID: lid, Hash: leaf.Hash})
+	}
+	return nodes, nil
+}
+
+// tileUpdateFn merges the base tile from the original map with the deltas that
+// represent the changes to the map. Note this only support additions or
```
Comment (AlCutter) [inline experimental/batchmap/tmap.go:193]: nit: supports

**Review by AlCutter — APPROVED** [review]
A few minor nits in-line

File: experimental/batchmap/README.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+Generates [Verifiable Maps](../../docs/papers/VerifiableDataStructures.pdf)
+using [Beam Go](https://beam.apache.org/get-started/quickstart-go/).
+Generating a map in batch scales better than incremental for large numbers of
+key/values.
+**This code is experimental**.
+
+The resulting map is output as tiles, in which the tree is divided from the
+root in a configurable number of prefix strata.
+Each strata contains a single byte of the 256-bit path.
+Tiles are only ouput if they are non-empty.
```
[the remark this answers is not in the document]
Comment (mhutchinson) [reply experimental/batchmap/README.md:12]: The plan is to launch and iterate on the docs. There is quite a bit of doc on the library functions themselves so I don't feel _too_ bad at the moment.

File: experimental/batchmap/tilepb/tile.proto (hunk 2 above)
Comment (mhutchinson) [reply experimental/batchmap/tilepb/tile.proto:33]: "We" don't do this, where "we" are the library maintainers. The `Entry` proto is provided as input to the map by the client. But I've extended this to say it should be obtained using a crypto hash.

**Comment by AlCutter** [comment]
LGTM, but linter is complaining about missing license header on a couple of files.

**Review by AlCutter — APPROVED** [review]

**Pull request merged**